### PR TITLE
Allow assign to workspace by number

### LINKED ISSUE #2590: Cannot assign to workspace by number
Output of `i3 --moreversion 2>&- || i3 --version`:

```
Binary i3 version:  4.13 (2016-11-08) © 2009 Michael Stapelberg and contributors
Running i3 version: 4.13 (2016-11-08) (pid 20337) abort…)
Loaded i3 config: /home/tarcisioe/.config/i3/config (Last modified: Tue 29 Nov 2016 04:35:01 BRST, 4838 seconds ago)

The i3 binary you just called: /usr/bin/i3
The i3 binary you are running: i3
```

I am attaching no log since this is a feature request. For consistency and expected behaviour, I believe `assign [<criteria>] workspace number <n>` should work in the same manner as `move to workspace number <n>` does. It right now actually assigns the window to a workspace named `number <n>`.

The workaround I have found is using `for_window [<criteria>] move to workspace number <n>`, but that has slightly different behaviour (for example, if I have moved a window out of its assigned workspace and issue a `i3 restart` command it is moved back to its assigned workspace. Using `assign` does not do that.)

To illustrate with my use case: I am using a small daemon with the ipc to dynamically rename my workspaces based on their contents. Therefore, for assigning windows to specific numbered workspaces I cannot use their full name, since it is dynamic.

Thanks in advance for the attention.

### PR DESCRIPTION
**Pull request by orestisf1993**
~~I haven't properly tested this, just a WIP take on the problem~~.

fixes #2590

### CODE REVIEW

**Comment by orestisf1993** [comment]
I've been using my branch for the last day and the change seems to work as intended. Does the "waiting" label refer to testcases?

**Comment by Airblader** [comment]
It refers to the fact that you called it a WIP. :-)

**Comment by orestisf1993** [comment]
I have added some testcases. I can also "dry" the testcase code a bit, commiting now.

**Comment by orestisf1993** [comment]
And can also merge the 2 cfg_assign functions

**Comment by orestisf1993** [comment]
I don't know why the `travis/check-spelling.pl` script fails and I can't run it locally.

**Comment by orestisf1993** [comment]
Just to be clear, I don't consider this a WIP anymore. I am pretty confident that the feature is complete functionality-wise.

**Comment by Airblader** [comment]
Yes, we'll just need to review it and get the tests running. I haven't had time to do so yet. I'll remove the label, though.

File: src/manage.c [hunk 1]
```diff
@@ -257,11 +257,21 @@ void manage_window(xcb_window_t window, xcb_get_window_attributes_cookie_t cooki
...
-        if ((assignment = assignment_for(cwindow, A_TO_WORKSPACE))) {
+        if ((assignment = assignment_for(cwindow, A_TO_WORKSPACE_NUMBER))) {
+            Con *output = NULL;
+            long parsed_num = ws_name_to_number(assignment->dest.workspace);
+
+            TAILQ_FOREACH(output, &(croot->nodes_head), nodes)
+            GREP_FIRST(assigned_ws, output_get_content(output),
+                       child->num == parsed_num);
+        }
+        if (assignment || (assignment = assignment_for(cwindow, A_TO_WORKSPACE))) {
```
Comment (Airblader) [inline src/manage.c]: I don't really like the flow of logic here. I think it'd read better like this:

```
if ((assignment = assignment_for(cwindow, A_TO_WORKSPACE)
    || (assignment = assignment_for(cwindow, A_TO_WORKSPACE_NUMBER))) {

    Con *assigned_ws = NULL;
    if (assignment->type == A_TO_WORKSPACE) {
        /* … */
        assigned_ws = …;
    } else {
        /* … */
        assigned_ws = …;
    }

    nc = …
}
```

File: src/manage.c [hunk 2]
```diff
@@ -257,11 +257,21 @@ void manage_window(xcb_window_t window, xcb_get_window_attributes_cookie_t cooki
...
+        Con *assigned_ws = NULL;
 
         /* If not, check if it is assigned to a specific workspace */
-        if ((assignment = assignment_for(cwindow, A_TO_WORKSPACE))) {
+        if ((assignment = assignment_for(cwindow, A_TO_WORKSPACE_NUMBER))) {
+            Con *output = NULL;
+            long parsed_num = ws_name_to_number(assignment->dest.workspace);
+
+            TAILQ_FOREACH(output, &(croot->nodes_head), nodes)
+            GREP_FIRST(assigned_ws, output_get_content(output),
```
Comment (Airblader) [inline src/manage.c]: I think this fits easily on a single line.

File: src/manage.c [hunk 3]
```diff
@@ -257,11 +257,21 @@ void manage_window(xcb_window_t window, xcb_get_window_attributes_cookie_t cooki
...
         Con *wm_desktop_ws = NULL;
+        Con *assigned_ws = NULL;
 
         /* If not, check if it is assigned to a specific workspace */
-        if ((assignment = assignment_for(cwindow, A_TO_WORKSPACE))) {
+        if ((assignment = assignment_for(cwindow, A_TO_WORKSPACE_NUMBER))) {
+            Con *output = NULL;
+            long parsed_num = ws_name_to_number(assignment->dest.workspace);
+
+            TAILQ_FOREACH(output, &(croot->nodes_head), nodes)
```
Comment (Airblader) [inline src/manage.c]: Please add `{}` around this for better readability and correct indentation.

**Comment by orestisf1993** [comment]
Done

**Pull request merged**